### PR TITLE
Add ignore_missing_files Hive config

### PR DESCRIPTION
**Pull request by zhli1142015**
Spark provides support for the data source configuration 
`spark.sql.files.ignoreMissingFiles`, to ignore missing files while reading data 
from files. 
https://spark.apache.org/docs/latest/sql-data-sources-generic-options.html#ignore-missing-files

Introduce Hive config ignore_missing_files to match this behavior.

### CODE REVIEW

**Comment by zhli1142015** [comment]
@rui-mo and @mbasmanova , could you help review this PR?
Thanks.

File: velox/connectors/hive/SplitReader.cpp [hunk 1]
```diff
@@ -91,7 +92,19 @@ void SplitReader::prepareSplit(
   VELOX_CHECK_NE(
       baseReaderOpts_.getFileFormat(), dwio::common::FileFormat::UNKNOWN);
 
-  auto fileHandle = fileHandleFactory_->generate(hiveSplit_->filePath).second;
+  std::shared_ptr<FileHandle> fileHandle = nullptr;
+  try {
+    fileHandle = fileHandleFactory_->generate(hiveSplit_->filePath).second;
+  } catch (VeloxRuntimeError& e) {
```
Comment (rui-mo) [inline velox/connectors/hive/SplitReader.cpp:98]: I have seen some discussions about the poor performance of try-catch, and am not sure if there's better way to achieve this. Maybe [Status](https://github.com/facebookincubator/velox/blob/main/velox/common/base/Status.h) is an option. Thanks.

File: velox/common/file/File.cpp [hunk 2]
```diff
@@ -103,13 +103,17 @@ uint64_t InMemoryWriteFile::size() const {
...
-  VELOX_CHECK_GE(
-      fd_,
-      0,
-      "open failure in LocalReadFile constructor, {} {} {}.",
-      fd_,
-      path,
-      folly::errnoStr(errno));
+  if (fd_ < 0) {
+    if (errno == ENOENT) {
+      VELOX_FILE_NOT_FOUND_ERROR("file {} does not exist", path);
```
Comment (rui-mo) [inline velox/common/file/File.cpp:108]: nit: messages ends with dot .

File: velox/common/file/tests/FileTest.cpp [hunk 3]
```diff
@@ -316,3 +316,18 @@ TEST(LocalFile, rmdir) {
   // deleted, which is not an error.
   EXPECT_NO_THROW(localFs->rmdir(tempFolder->path));
 }
+
+TEST(LocalFile, fileNotFound) {
+  filesystems::registerLocalFileSystem();
+  auto tempFolder = ::exec::test::TempDirectoryPath::create();
+  auto path = fmt::format("{}/file", tempFolder->path);
+  auto localFs = filesystems::getFileSystem(path, nullptr);
+  {
```
Comment (mbasmanova) [inline velox/common/file/tests/FileTest.cpp:325]: nit: remove this scope

File: velox/common/file/tests/FileTest.cpp [hunk 4]
```diff
@@ -316,3 +316,18 @@ TEST(LocalFile, rmdir) {
...
+    try {
+      LocalReadFile readFile(path);
+      ASSERT_FALSE(true) << "Function should throw.";
```
Comment (mbasmanova) [inline velox/common/file/tests/FileTest.cpp:328]: FAIL()

File: velox/connectors/hive/HiveConfig.cpp [hunk 5]
```diff
@@ -152,6 +152,14 @@ bool HiveConfig::isPartitionPathAsLowerCase(const Config* session) const {
   return session->get<bool>(kPartitionPathAsLowerCaseSession, true);
 }
 
+bool HiveConfig::ignoreMissingFiles(const Config* session) const {
+  if (session->isValueExists(kIgnoreMissingFilesSession)) {
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.cpp:156]: This pattern of double checking (exists + get) is an anti-pattern. Let's change to `session->get<T>(name, defaultValue)`. Would you make a separate PR for that? CC: @kewang1024

File: velox/connectors/hive/SplitReader.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/connectors/hive/SplitReader.cpp:98]: @rui-mo Throwing is problematic when it happens per-row. Here, throwing should be fine.

File: velox/connectors/hive/SplitReader.cpp [hunk 6]
```diff
@@ -91,7 +92,19 @@ void SplitReader::prepareSplit(
   VELOX_CHECK_NE(
       baseReaderOpts_.getFileFormat(), dwio::common::FileFormat::UNKNOWN);
 
-  auto fileHandle = fileHandleFactory_->generate(hiveSplit_->filePath).second;
+  std::shared_ptr<FileHandle> fileHandle = nullptr;
```
Comment (mbasmanova) [inline velox/connectors/hive/SplitReader.cpp:95]: drop  = nullptr

File: velox/docs/configs.rst [hunk 7]
```diff
@@ -424,6 +424,11 @@ Each query can override the config by setting corresponding query session proper
      - bool
      - true
      - If true, the partition directory will be converted to lowercase when executing a table write operation.
+   * - ignore_missing_files
+     -
+     - bool
+     - false
+     - If true, the task will continue to run when encountering missing files and the split woule be marked as empty split.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:431]: If true, splits that refer to missing files don't generate errors and are processed as empty splits.

File: velox/exec/tests/TableScanTest.cpp [hunk 8]
```diff
@@ -1388,12 +1388,28 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
-  CursorParameters params;
-  params.planNode = tableScanNode();
-
-  auto cursor = TaskCursor::create(params);
-  cursor->task()->addSplit("0", makeHiveSplit("/path/to/nowhere.orc"));
-  EXPECT_THROW(cursor->moveNext(), VeloxRuntimeError);
+  auto split = HiveConnectorSplitBuilder("/path/to/nowhere.orc").build();
+  for (bool ignoreMissingFiles : {true, false}) {
+    auto assertQueryForMissingFile = [&](bool ignoreMissingFile) {
+      AssertQueryBuilder(tableScanNode(), duckDbQueryRunner_)
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1394]: remove duckDbQueryRunner_

File: velox/exec/tests/TableScanTest.cpp [hunk 9]
```diff
@@ -1388,12 +1388,28 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
+          .connectorSessionProperty(
+              kHiveConnectorId,
+              connector::hive::HiveConfig::kIgnoreMissingFilesSession,
+              std::to_string(ignoreMissingFile))
+          .splits({split})
+          .assertResults("");
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1400]: use assertEmptyResults

File: velox/exec/tests/TableScanTest.cpp [hunk 10]
```diff
@@ -1388,12 +1388,28 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
-  EXPECT_THROW(cursor->moveNext(), VeloxRuntimeError);
+  auto split = HiveConnectorSplitBuilder("/path/to/nowhere.orc").build();
+  for (bool ignoreMissingFiles : {true, false}) {
+    auto assertQueryForMissingFile = [&](bool ignoreMissingFile) {
+      AssertQueryBuilder(tableScanNode(), duckDbQueryRunner_)
+          .connectorSessionProperty(
+              kHiveConnectorId,
+              connector::hive::HiveConfig::kIgnoreMissingFilesSession,
+              std::to_string(ignoreMissingFile))
+          .splits({split})
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1399]: use split(split)

File: velox/exec/tests/TableScanTest.cpp [hunk 11]
```diff
@@ -1388,12 +1388,28 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
 TEST_F(TableScanTest, fileNotFound) {
-  CursorParameters params;
-  params.planNode = tableScanNode();
-
-  auto cursor = TaskCursor::create(params);
-  cursor->task()->addSplit("0", makeHiveSplit("/path/to/nowhere.orc"));
-  EXPECT_THROW(cursor->moveNext(), VeloxRuntimeError);
+  auto split = HiveConnectorSplitBuilder("/path/to/nowhere.orc").build();
+  for (bool ignoreMissingFiles : {true, false}) {
+    auto assertQueryForMissingFile = [&](bool ignoreMissingFile) {
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1393]: auto assertMissingFile = .

File: velox/exec/tests/TableScanTest.cpp [hunk 12]
```diff
@@ -1388,12 +1388,28 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
+              connector::hive::HiveConfig::kIgnoreMissingFilesSession,
+              std::to_string(ignoreMissingFile))
+          .splits({split})
+          .assertResults("");
+    };
+    if (ignoreMissingFiles) {
+      assertQueryForMissingFile(ignoreMissingFiles);
+    } else {
+      try {
+        assertQueryForMissingFile(ignoreMissingFiles);
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1406]: use VELOX_ASSERT_THROW

File: velox/common/file/tests/FileTest.cpp [hunk 13]
```diff
@@ -316,3 +316,18 @@ TEST(LocalFile, rmdir) {
...
+    } catch (VeloxRuntimeError& e) {
+      ASSERT_EQ(e.errorCode(), error_code::kFileNotFound);
```
Comment (mbasmanova) [inline velox/common/file/tests/FileTest.cpp:330]: Let's add a macro similar to VELOX_ASSERT_THROW that can check error code as well.

**Review by mbasmanova — COMMENTED** [review]
@zhli1142015 Looks good overall. Some comments.

File: velox/connectors/hive/HiveConfig.cpp (hunk 5 above)
Comment (zhli1142015) [reply velox/connectors/hive/HiveConfig.cpp]: Updated, thanks.

File: velox/connectors/hive/SplitReader.cpp (hunk 6 above)
Comment (zhli1142015) [reply velox/connectors/hive/SplitReader.cpp]: Updated, thanks.

File: velox/docs/configs.rst (hunk 7 above)
Comment (zhli1142015) [reply velox/docs/configs.rst]: Updated, thanks.

File: velox/exec/tests/TableScanTest.cpp (hunk 8 above)
Comment (zhli1142015) [reply velox/exec/tests/TableScanTest.cpp]: Updated, thanks.

File: velox/exec/tests/TableScanTest.cpp (hunk 9 above)
Comment (zhli1142015) [reply velox/exec/tests/TableScanTest.cpp]: Updated, thanks.

File: velox/exec/tests/TableScanTest.cpp (hunk 12 above)
Comment (zhli1142015) [reply velox/exec/tests/TableScanTest.cpp]: Added, thanks.

File: velox/common/file/tests/FileTest.cpp (hunk 13 above)
Comment (zhli1142015) [reply velox/common/file/tests/FileTest.cpp]: Added, thanks.

File: velox/connectors/hive/HiveConfig.cpp (hunk 5 above)
Comment (zhli1142015) [reply velox/connectors/hive/HiveConfig.cpp]: Sure.

Comment (zhli1142015) [reply velox/connectors/hive/HiveConfig.cpp]: https://github.com/facebookincubator/velox/pull/8633, thanks.

File: velox/common/base/tests/GTestUtils.h [hunk 14]
```diff
@@ -63,6 +63,16 @@
       << "Expected error message to contain '" << (_errorMessage)          \
       << "', but received '" << status.message() << "'."
 
+#define VELOX_ASSERT_ERROR_CODE(_expression, _type, _errorCode)               \
```
Comment (mbasmanova) [inline velox/common/base/tests/GTestUtils.h:66]: nit: perhaps, VELOX_ASSERT_RUNTIME_THROW_CODE(_expression, _errorCode) to match VELOX_ASSERT_RUNTIME_THROW

File: velox/exec/tests/TableScanTest.cpp [hunk 15]
```diff
@@ -1388,12 +1388,26 @@ TEST_F(TableScanTest, splitOffsetAndLength) {
...
 
 TEST_F(TableScanTest, fileNotFound) {
-  CursorParameters params;
-  params.planNode = tableScanNode();
-
-  auto cursor = TaskCursor::create(params);
-  cursor->task()->addSplit("0", makeHiveSplit("/path/to/nowhere.orc"));
-  EXPECT_THROW(cursor->moveNext(), VeloxRuntimeError);
+  auto split = HiveConnectorSplitBuilder("/path/to/nowhere.orc").build();
+  for (bool ignoreMissingFiles : {true, false}) {
```
Comment (mbasmanova) [inline velox/exec/tests/TableScanTest.cpp:1392]: nit: perhaps, it would be more readable as 

```
assertMissingFile(true);

VELOX_ASSERT_RUNTIME_CODE_THROW(
          assertMissingFile(false),
          error_code::kFileNotFound);
```

without the loop

**Review by mbasmanova — COMMENTED** [review]
Looks great % a couple nits.

File: velox/common/base/tests/GTestUtils.h (hunk 14 above)
Comment (zhli1142015) [reply velox/common/base/tests/GTestUtils.h]: Added, thanks.

File: velox/exec/tests/TableScanTest.cpp (hunk 15 above)
Comment (zhli1142015) [reply velox/exec/tests/TableScanTest.cpp]: Updated, thanks.

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: velox/connectors/hive/SplitReader.cpp [hunk 16]
```diff
@@ -91,7 +92,19 @@ void SplitReader::prepareSplit(
   VELOX_CHECK_NE(
       baseReaderOpts_.getFileFormat(), dwio::common::FileFormat::UNKNOWN);
 
-  auto fileHandle = fileHandleFactory_->generate(hiveSplit_->filePath).second;
+  std::shared_ptr<FileHandle> fileHandle;
+  try {
+    fileHandle = fileHandleFactory_->generate(hiveSplit_->filePath).second;
```
Comment (mbasmanova) [inline velox/connectors/hive/SplitReader.cpp:97]: Would you also document 'generate' API to explain that it should throw a specific error code to indicate file-not-found?

Comment (zhli1142015) [reply velox/connectors/hive/SplitReader.cpp:97]: I plan to add doc for this in the follow up pr, sth like: https://github.com/facebookincubator/velox/pull/8645/commits/c195d903fedfeaded65ee0a25b0874bc1321973d#diff-b2b8b019f80c44db000432c6c90bfffeb215c772b15908246f736ddd30011d1eR84. Please let me know if other docs are needed, thanks.

**Review by mbasmanova — COMMENTED** [review]
@zhli1142015 This change is breaking Prestissimo tests.

```
[ RUN      ] BroadcastTest.invalidBroadcastFilePath
fbcode/github/presto-trunk/presto-native-execution/presto_cpp/main/operators/tests/BroadcastTest.cpp:339: Failure
Value of: e.message().find("No such file or directory") != std::string::npos
  Actual: false
Expected: true
Expected error message to contain 'No such file or directory', but received 'file /tmp/this-should-not-exist/velox--missing-broadcast-file.bin does not exist.'.
```

Maybe change

```
VELOX_FILE_NOT_FOUND_ERROR("file {} does not exist.", path);
```

to

```
VELOX_FILE_NOT_FOUND_ERROR("No such file or directory: {}", path);
```

**Comment by zhli1142015** [comment]
> @zhli1142015 This change is breaking Prestissimo tests.
> 
> ```
> [ RUN      ] BroadcastTest.invalidBroadcastFilePath
> fbcode/github/presto-trunk/presto-native-execution/presto_cpp/main/operators/tests/BroadcastTest.cpp:339: Failure
> Value of: e.message().find("No such file or directory") != std::string::npos
>   Actual: false
> Expected: true
> Expected error message to contain 'No such file or directory', but received 'file /tmp/this-should-not-exist/velox--missing-broadcast-file.bin does not exist.'.
> ```
> 
> Maybe change
> 
> ```
> VELOX_FILE_NOT_FOUND_ERROR("file {} does not exist.", path);
> ```
> 
> to
> 
> ```
> VELOX_FILE_NOT_FOUND_ERROR("No such file or directory: {}", path);
> ```

Updated, thanks.
